foyer/general_forcefield.py

```diff
@@ -0,0 +1,539 @@
+from gmso.core import element
+
+import mbuild as mb
+from foyer.atomtyper import find_atomtypes
+from foyer.exceptions import FoyerError
+from foyer import smarts
+from foyer.validator import Validator
+from foyer.xml_writer import write_foyer
+from foyer.utils.io import import_, has_mbuild
+from foyer.utils.external import get_ref
+
+
+# Copy from original forcefield.py
+def preprocess_forcefield_files(forcefield_files=None, backend='gmso'):
+    """Pre-process foyer Forcefield XML files"""
+    if forcefield_files is None:
+        return None
+
+    preprocessed_files = []
+
+    # Attempt to preprocessed the the xml files provided and make sure
+    # that they are in compliance with the foyer format
+    for xml_file in forcefield_files:
+        if not hasattr(xml_file, 'read'):
+            f = open(xml_file)
+            _, suffix = os.path.split(xml_file)
+        else:
+            f = xml_file
+            suffix = ""
+
+        # read and preprocess
+        xml_contents = f.read()
+        f.close()
+        xml_contents = re.sub(r"(def\w*=\w*[\"\'])(.*)([\"\'])", lambda m: m.group(1) + re.sub(r"&(?!amp;)", r"&amp;", m.group(2)) + m.group(3),
+                              xml_contents)
+
+        try:
+            '''
+            Sort topology objects by precedence, defined by the number of
+            `type` attributes specified, where a `type` attribute indicates
+            increased specificity as opposed to use of `class`
+            '''
+            root = ET.fromstring(xml_contents)
+            for element in root:
+                if 'Force' in element.tag:
+                    element[:] = sorted(element, key=lambda child: (
+                        -1 * len([attr_name for attr_name in child.keys()
+                                    if 'type' in attr_name])))
+            xml_contents = ET.tostring(root, method='xml').decode()
+        except ET.ParseError:
+            '''
+            Provide the user with a warning if sorting could not be performed.
+            This indicates a bad XML file, which will be passed on to the
+            Validator to yield a more descriptive error message.
+            '''
+            warnings.warn('Invalid XML detected. Could not auto-sort topology '
+                          'objects by precedence.')
+
+
+    tmp_processed_files = list()
+    if backend == 'gmso':
+        # Run through the forcefield XML conversion
+        from gmso.external.convert_foyer_xml import from_foyer_xml
+        for idx, file in enumerate(forcefield_files):
+            _, suffix = os.path.split(file)
+            tempfile = NamedTemporaryFile(suffix=suffix, delete=False)
+            from_foyer_xml(foyer_xml=str(file),
+                          gmso_xml=str(tempfile.name), overwrite=True)
+            tmp_processed_files.append(tempfile.name)
+
+    else:
+        raise FoyerError('Backend not supported')
+
+    return tmp_processed_files
+
+
+class Forcefield(object):
```

Comment:
This is a general comment to think about for the development process.  

When Umesh was reviewing my GOMC PR, the forcefield_files and name were in a separate selection.  In the GOMC case, the FFs are set as a single string (covers all molecules) or by a specific residue selection (which is molecule selection) via a dictionary ({Residue_name: forcefield_file or name}).  

This approach combines the forcefield_files and name into 1 variable and allows the selection of an existing/standard foyer package and selecting the many forcefield_files.  This method ensures or at least helps minimize overlapping molecule types between the selected FFs, which may be difficult for some new users.

I'm not sure it would work directly without changing a lot of things.  However, for me and the GOMC code, it did simplify it quite a bit.  This was circumvented by pre-processing the molecules into residue/molecule bins before the FF process.  In the GOMC case, we can use this dictionary to utilize both the standard foyer FF and add other xml FF files.  For example: 
 -The selection based on residues/molecule type also allows the user to add a different water FF than the existing one (say TIP3P instead of TIP4P for the oplsaa FF), which simplifies the user input to creating a TIP3P xml, rather than messing with the oplsaa FF and risk messing it up. 

The GOMC case likely will be covered in the way listed above either way, but it is just a general comment.  Again, not saying it needs to be fixed in this PR.